### DISCUSSION TITLE
Timeout parameter when writing cells?

### DISCUSSION THREAD

**Discussion by raeldor** [cubewise-code/tm1py#981] [General]
Is it possible to specify a timeout when writing cells using write_values?  Thanks.

**Comment by MariusWirtz** [comment]
Yes. You can pass `timeout` and `cancel_at_timeout` to any function in TM1py.

However, if you use the `write` function with `use_blob=True` it is so much faster that you might not need the timeout.

**Comment by raeldor** [comment]
I think our forked version might be too old then, as I don't see that parameter for timeout.  Maybe I'll try and merge out get_values() in to the newest codebase.

**Comment by raeldor** [comment]
I changed the write_values() to write() sucessfully, but adding the use_blob=True results in 'the JSON object must be str, bytes or bytearray, not list' error.

**Comment by raeldor** [comment]
I also tried adding the timeout parameters, but it's still not timing out. :(

**Comment by raeldor** [comment]
Timeout parameters don't seem to have any effect, but I used write() where I could.  However it seems write() doesn't always adhere to the same rules where I can use 'P1000.0' to push to consolidated level.  Somehow with write() the 'P1000.0' gets through to TM1 and  I get an error about conversion to numeric.  Doesn't happen with write_values()

**Reply by MariusWirtz** [reply]
I'm working on the master branch. The `timeout` and `cancel_at_timeout` arguments make a difference for me.

``` python
from TM1py.Services.TM1Service import TM1Service

tm1_params = {...}

with TM1Service(**tm1_params) as tm1:
    cells = dict()
    for i in range(1, 11):
        for j in range(1, 11):
            cells[f"e{i}", f"e{j}"] = i * j

    tm1.cells.write_values("c1", cells, timeout=0.1, cancel_at_timeout=True)

```

``` python
from TM1py.Services.TM1Service import TM1Service

tm1_params = {...}

with TM1Service(**tm1_params) as tm1:
    cells = dict()
    for i in range(1, 11):
        for j in range(1, 11):
            cells[f"e{i}", f"e{j}"] = i * j

    tm1.cells.write_values("c1", cells)

```

Indeed, the `write` function doesn't support the spreading commands. It's because we optimized it for write-speed and don't use the default REST API write operation.
Only the `write_values` function (which uses the "normal" write operation in the REST API) supports the spreading commands.

**Comment by raeldor** [comment]
Thank you for the reply.  I had assumed the timeout was in seconds, but having seen your code has '0.1', now I'm wondering whether it is minutes?

**Comment by MariusWirtz** [comment]
It is measured in seconds but you can pass a  float
